Celeste.Mod.mm/Patches/Monocle/Tracker.cs

```diff
@@ -56,72 +67,115 @@ private static List<Type> GetSubclasses(Type type) {
+            bool updated = false;
+            // copy the registered types for the target type
+            ((bool) trackedEntity ? StoredEntityTypes : StoredComponentTypes).Add(type);
+            Dictionary<Type, List<Type>> tracked = (bool) trackedEntity ? TrackedEntityTypes : TrackedComponentTypes;
+            if (!type.IsAbstract) {
+                if (!tracked.TryGetValue(type, out List<Type> value)) {
+                    value = new List<Type>();
+                    tracked.Add(type, value);
+                }
+                int cnt = value.Count;
+                value.AddRange(tracked.TryGetValue(trackedAsType, out List<Type> list) ? list : new List<Type>());
+                List<Type> result = tracked[type] = value.Distinct().ToList();
+                updated = cnt != result.Count;
+            }
+            // do the same for subclasses
+            foreach (Type subtype in subtypes) {
+                if (trackedAsType.IsAssignableFrom(subtype) && !subtype.IsAbstract) {
+                    if (!tracked.TryGetValue(subtype, out List<Type> value)) {
+                        value = new List<Type>();
+                        tracked.Add(subtype, value);
+                    }
+                    int cnt = value.Count;
+                    value.AddRange(tracked.TryGetValue(trackedAsType, out List<Type> list) ? list : new List<Type>());
+                    List<Type> result = tracked[subtype] = value.Distinct().ToList();
+                    updated = cnt != result.Count;
+                }
+            }
+            if (updated) {
+                TrackedTypeVersion++;
+            }
+        }
+
+        /// <summary>
+        /// Ensures the current scene's tracker contains all entities of all tracked Types.
+        /// Must be called if a type is added to the tracker manually and if the active scene changes.
+        /// If called back to back without a type added to the Tracker, it won't go through again, for performance.
+        /// <paramref name="force"/> will make ensure the Refresh happens, even if run back to back.
+        /// Due to only the active Scene's Tracker's refreshed state is changed, <paramref name="force"/> must be true if a different scene becomes active.
+        /// </summary>
+        public void Refresh(bool force = false) {
+            if (currentVersion >= TrackedTypeVersion && !force) {
+                return;
+            }
+            currentVersion = TrackedTypeVersion;
+            foreach (Type entityType in StoredEntityTypes) {
+                if (!Entities.ContainsKey(entityType)) {
+                    Entities.Add(entityType, new List<Entity>());
+                }
+            }
+            foreach (Type componentType in StoredComponentTypes) {
+                if (!Components.ContainsKey(componentType)) {
+                    Components.Add(componentType, new List<Component>());
+                }
+            }
+            foreach (Entity entity in Engine.Scene.Entities) {
```

Comment:
You are using `Engine.Scene` on an instance method of a `Tracker`.
There is no guarantee that `Engine.Scene.Tracker == this`, which is bad for this usecase.
So the options are either adding a `Scene` field and setting that from the constructor, or having a `Scene` parameter which the user needs to pass in.